### ISSUE TITLE
Not able to see the first page of my app. 

### ISSUE THREAD

**Issue by jeeva-nikam** [apache/cordova-plugin-wkwebview-engine#166]
# Bug Report

### What does actually happen?
I am using Angularjs, in that <div ng-view></div> is not loading the default login page route.
App is loading the index.html only problem with routing.  

### Version information

*Cordova: Cordova CLI - 8.0.0, 
*Cordova Platforms iOS - 4.4.0, 
*Cordova Plugins - 
cordova-plugin-actionsheet 2.3.1 "ActionSheet"
cordova-plugin-battery-status 2.0.2 "Battery"
cordova-plugin-camera 2.1.1 "Camera"
cordova-plugin-compat 1.0.0 "Compat"
cordova-plugin-console 1.0.4 "Console"
cordova-plugin-contacts 2.0.1 "Contacts"
cordova-plugin-datepicker 0.9.3 "DatePicker"
cordova-plugin-device 1.1.3 "Device"
cordova-plugin-device-motion 1.2.2 "Device Motion"
cordova-plugin-device-orientation 1.0.4 "Device Orientation"
cordova-plugin-dialogs 1.2.1 "Notification"
cordova-plugin-file 4.3.0 "File"
cordova-plugin-file-transfer 1.5.1 "File Transfer"
cordova-plugin-flashlight 3.2.0 "Flashlight"
cordova-plugin-geolocation 2.1.0 "Geolocation"
cordova-plugin-globalization 1.0.4 "Globalization"
cordova-plugin-inappbrowser 1.3.0 "InAppBrowser"
cordova-plugin-mauron85-background-geolocation 2.3.6 "CDVBackgroundGeolocation"
cordova-plugin-media 2.2.0 "Media"
cordova-plugin-media-capture 1.2.0 "Capture"
cordova-plugin-network-information 1.2.1 "Network Information"
cordova-plugin-splashscreen 3.2.2 "Splashscreen"
cordova-plugin-statusbar 2.1.3 "StatusBar"
cordova-plugin-vibration 2.1.2 "Vibration"
cordova-plugin-whitelist 1.2.2 "Whitelist"
cordova-plugin-wkwebview-engine 1.2.1 "Cordova WKWebView Engine"
phonegap-plugin-barcodescanner 6.0.2 "BarcodeScanner"
phonegap-plugin-push 1.8.2 "PushPlugin"
uk.co.workingedge.phonegap.plugin.launchnavigator 3.2.0 "Launch Navigator"

*Other Frameworks: Ionic Framework and CLI version
*Operating System - Xcode etc.

## Checklist

- [ ] I searched for existing GitHub issues
- [ ] I updated all Cordova tooling to most recent version
- [ ] I included all the necessary information above

**Comment by dec** [comment]
Hello to all,

I am able to run an AngularJS app working by installtion the "cordova-plugin-wkwebview-engine" plugin but also the "https://github.com/TheMattRay/cordova-plugin-wkwebviewxhrfix" plugin. Maybe you want to try it.

Additionally, maybe someone read this... and want to answer this question: why we must use the referred "fix" plugin in addition to the "cordova-plugin-wkwebview-engine" plugin? I mean, the "fix" plugin allows us to run apps made by AngularJS but also VueJS, for example.

Without the "fix" plugin we can't see our apps working, so, why not to apply that fix in the "cordova-plugin-wkwebview-engine" itself? It's better to use another plugin (like the linked one) in order to fix the problem with XHR requests for local files? I appreciate if someone can give me some light about this question.

Maybe I can open another issue to post this question?

**Comment by breautek** [comment]
To answer this question:

You are likely running into a CORS problem when requesting local files. You'll need a wkwebview file xhr plugin.

Alternatively, you can upgrade to `cordova-ios@6`, remove `cordova-plugin-wkwebview` and use schemes. This will however cause you to lose web storage data. The data will still be on the device, but it will not be accessible, because of the origin change.

> Additionally, maybe someone read this... and want to answer this question: why we must use the referred "fix" plugin in addition to the "cordova-plugin-wkwebview-engine" plugin? I mean, the "fix" plugin allows us to run apps made by AngularJS but also VueJS, for example.

### Little background

The reason why the wkwebview xhr fix is necessary for apps that uses certain frameworks like Angular is because those frameworks makes heavy usage of `XMLHttpRequest` to local urls (`file://`...), which is illegal in the webview because of CORS. That fix plugin works by overwriting the webviews version of XMLHttpRequest with a native version that doesn't enforce CORS on local urls.

I don't know if there was ever a community decision but personally I have a few problems with approach:

1. Overwriting native behaviour is generally frowned upon by Apple, especially when it comes to security constructs. While I never heard of an app being rejected because of the wkwebview xhr fix, I would be concerned that it is possible.

2. Generally if you need to read files from the local file system, you won't go through a network stack, you'd simply read the files directly using the file APIs. Angular framework obviously makes certain assumptions, and one of those assumptions is that it is running on a website, powered by a webserver. Which is a false assumption with Cordova apps.

### Cordova's approach

Cordova has chosen a different approach to tackle this problem, which is available in cordova-ios@6. If you're using schemes, you don't need the wkwebview xhr fix plugin. Because schemes have an origin attached which eliminates several CORS-related issues, not just the xhr issue.

This still does pose a different problem however, for existing apps. Because changing from `file://` to a custom URL scheme changes the origin, you'll lose acccess to any data stored in web storage (local storage, indexedDB, cookies, etc). Because of this, cordova chose to keep the `file://` protocol as a default in cordova-ios@6, but it does break the wkwebview file xhr plugin.

side note: the plugin you linked to isn't being maintained. There is a plugin maintained by Oracle that I would probably suggest https://github.com/oracle/cordova-plugin-wkwebview-file-xhr to use instead, and they are working on upgrading their plugin to support cordova-ios@6.

**Comment by dec** [comment]
Hello to all,

Many, many thanks for your explendid answer, @breautek, I really appreciate it. After you point me in the right direction, I read this [post in the Apache Cordova's blog](https://cordova.apache.org/announcements/2020/06/01/cordova-ios-release-6.0.0.html), since, apparently with cordova-ios 6.0 we no need the referred plugins at all. 

Some one in the referred blog post, however, apparently have problems by using scheme's options like the below one:

`<preference name="scheme" value="file"/>
<preference name="hostname" value="localhost"/>`

Can you please, post here what are the correct values for "scheme" and "hostname" in order to be ready with AngularJS and VueJS apps, for example? Anyway, when possible, I want to test the referred scheme's options, since I must admit that I am not already try that.

Anyway, thanks again for your great and clarificate answer, @breautek.

**Comment by breautek** [comment]
> <preference name="scheme" value="file"/>

Scheme cannot be any well known/defined scheme on iOS. So you cannot use `file`, or `http`, `https`, etc.

Using `app` should work, it will make your urls look like `app://.....` but, the disclaimer still applies and your web storage will reset. For some people this isn't an issue. If web storage is very important to keep, you either need to find a way to migrate it using a plugin, or stick with the `file://` protocol.

**Comment by dec** [comment]
Hello to all,

Thanks again for your answer, @breautek. I am trying here with "scheme=app", and, unfortunately, I can't see an AngularJS app working, that is, what I get is a "white screen", and, if I am not wrong, this is related with the XHR usage by AngularJS.

I am wrong? That is, what I expected, is to view the app running in the same way that I can see it running using the "cordova-plugin-wkwebview-engine" and the "fix" plugin: without any change!

Because about the possible "app://" I am not quite sure if we must modify the relative paths in the images, for example, to start to use the "app://" protocol... and, if so,... how to tell to AngularJS (I think in VueJS there is the same problem) to follow that protocol instead of the "default" one...

I am working around and will post here if finally I find any possible solution, trying to use cordova-ios 6 and not the plugins, because what you post above: maybe right now there is not a problem to use the "fix" plugin, but, if we can do the things without plugins... probably it's the best.

Thanks again for your help, @breautek.

**Comment by dec** [comment]
Hello to all,

Just to add more information. Following the @breautek indications, that is, testing an AngularJS app with only the Cordova's splash screen plugin, cordova-ios 6.1.0 and the option scheme=app, I can't see it working as expected. After the splash screen what I can see is only a "white screen", probably because (99% sure) there is some problem related with AngularJS or the AngularJS route plugin, I am not sure at this time.

However, testing the same with an VueJS app with only the Cordova's splash screen plugin, cordova-ios 6.1.0 and the option scheme=app, I can see the app working as expected, and, even some images using relative paths, are shown also as expected without problem.

So apparently there is something related with AngularJS that I can't catch for now... and I must trying to do, so, if I can find any solution I will post it here to anyone who can be interested.

Thanks ver much again, @breautek, certainly the possible AngularJS issue still exists, but, since the VueJS app is working exactly as expected following your instructions, I think that there must be no reason to be not working with AngularJS, so, or I am wrong, or the issue is related with AngularJS and not with Apache Cordova.

**Comment by breautek** [comment]
Make sure you're using the latest version of the splashscreen plugin. Splashscreens are not considered part of the core platform, so the splashscreen plugin is also obsolete for ios. You still need the plugin for other platforms however at this time, but the latest release prevents it from being installed for the ios platform.

**Comment by lovelyelfpop** [comment]
How to display local images in application storage directory (e.g. file:///var/mobile/Applications//) after add <preference name="scheme" value="cdv" /> <preference name="hostname" value="localhost" /> ?

**Comment by dec** [comment]
Hello to all,

Nothing. I try several possible solutions, here just the ones that I remember... sorry to not post here everything what I am trying, but, maybe can be resumed in these ones:

1º Trying to adapt this [possible solution for Angular](https://stackoverflow.com/questions/59866439/why-angular-8-router-is-not-working-in-cordova-ios-with-wkwebview) into AngularJS, but can't get it working. I am not completely sure if I adapt well this possible solution from Angular to AngularJS, I think so, but can't be sure. Maybe there is a way in which XCode offer to us information about the JavaScript possible problems? Since the XCode console don't offer any kind of error message...

2º Trying to avoid the splash screen, not the plugin, but any kind of splash screen image to be used. Here what I can see (not related with the main problem) is that the preferences options to disabled the splash screen don't work: apparently since the config.xml file have some "storyboard" splash screen images, cordova-ios 6.1.0 pick it to be used. Anyway, I remove the images from the config.xml... so what I can see now is a white screen... not the splash screen... but a white screen...

So, if I am not wrong, the issue is related with AngularJS, maybe not by itself, maybe for some specific option / configuration / code usage of my specific app. However, I can figure out that I am not the only one with this issue (the white screen...)... as this issue open message apparently have the same issue, with the same framework (AngularJS).

I will try to continue investigating around, and, again, if finally I can find any solution I will post it here.

**Comment by breautek** [comment]
> Maybe there is a way in which XCode offer to us information about the JavaScript possible problems? Since the XCode console don't offer any kind of error message...

You can remotely debug your app using the safari web inspector. Note that errors are only recorded if the inspector is opened, so to see startup errors, you must do a page refresh, which you can do by pressing cmd + r while the web inspector is focused.

**Comment by dec** [comment]
Hello to all,

Many thanks @breautek, I will try that, becuase it's quite probable that the Safari web inspector show someting useful.

**Comment by dec** [comment]
Hello to all,

I must admit that the Safari web inspector is new to me, and, it's another thing to say thanks you, @breautek, because for sure this can be very useful.

Back into the issue, however... there is something rare here. To be sure that I configure well the Safari web inspector in both the Mac OS and my iPad, what I do is to try with the VueJS app, and, certainly, the web inspector is working providing very useful information.

When I start the VueJS based app from XCode, I can see in the "Development" menu of Safari an app, an entry for my device named "--localhost", so I just click it and the Safari web inspector become visible and working as expected.

However, when I try the same with the AngularJS app, the "Development" menu of Safari shown me a "about:blank" entry to be debugged... I click on the item, the web inspector appear, but, unfortunately there is no information... nothing at all... just an "about:blank" entry in the "Red tab" of the inspector, but nothing int he "Console tab" nor in any other tab.

I think I must stop for now... since I have all the day working in this issue, and, I am a bit tired now... but want to continue to working on this, since it's important to me to get my AngularJS app working as expected using cordova-ios 6.x.

Anyway, I want to say thanks again to you, @breautek, I learn several things thanks to you today: from the Safari web inspector possible usage (and I will do it for sure from today), to the ability to use the Apache Cordova 6.x instead of the "cordova-plugin-wkwebview-engine" and the "fix" plugins, which certainly works as expected in my VueJS app.

Many thanks for your amazing help that you provided, @breautek! Again, if I finally can find any possible solution I will post it here. And, also continue to take a look at this and other possible similar issues related with AngularJS and cordova-ios 6.x. Said this, as I mention before, maybe the problem is not AngularJS itself, but something in my specific app. It's rare, since other people have the same issue with AngularJS, but, maybe we are shared the same "wrong" configuration or something like that.

**Comment by dec** [comment]
Hello to all,

Ok, finally I got my AngularJS app working in cordova-ios 6.1.0. What to say... I am quite, quite stupid! Seeing that my VueJS app is working as expected, and, seeing that the "app://localhost" protocol is what is used after configured the "scheme" and the "hostname" with "app" and "localhost", respectively, I can't see a reason to not get working in the same way my AngularJS app.

So, after several tries... I take a look at the configuration, and what I see suprise me... I have defined the "hostname" in the below way:

`<preference name="hostname" value="localhost " />`

Note the space after "locahost"!! So of course I decide to remove that stupid space and try again and... the AngularJS app is working as expected now.

My recommendation for @jeeva-nikam is to follow the right instructions of @breautek, that is, directly use the cordova-ios 6x avoiding the usage of any plugin.

My gratitude to @breautek for his amazing help: he point me in the very right direction!
